### DISCUSSION TITLE
How should I approach formatting??

### DISCUSSION THREAD

**Discussion by idelice** [LazyVim/LazyVim#1965] [General]
Using jdtls and conform plugins, I'm trying to get my head around on how the heck I should configure formatting for specific file types - java in this case. 

Which of those plugins do I modify and what exactly do I need to modify in order to make it work?
How do I enable my own custom formatting from my a file?

**Comment by rogeriofrsouza** [comment]
This is the [documentation](https://www.lazyvim.org/plugins/formatting) on how to use conform.nvim
Here is a code that worked for me, [google-java-format](https://github.com/google/google-java-format) is shown here but you can use any [formatter supported](https://github.com/stevearc/conform.nvim?tab=readme-ov-file#formatters)
```lua
{
  "stevearc/conform.nvim",
  opts = {
    formatters_by_ft = {
        java = { "google-java-format" },
    }
  }
}
```

However, AFAIK google-java-format cannot be configured. That's why I use jdtls as my formatter through [nvim-jdtls](https://github.com/mfussenegger/nvim-jdtls)
This is how you can configure it with the [Java extra](https://www.lazyvim.org/extras/lang/java)
```lua
return {
  "mfussenegger/nvim-jdtls",
  opts = {
    jdtls = function(opts)
      opts.settings = {
        java = {
          format = {
            enabled = true,
            settings = {
              url = vim.fn.expand("~/Workspace/eclipse-java-google-style.xml"),
              profile = "GoogleStyle",
            },
          },
          inlayHints = {
            parameterNames = {
              enabled = "all",
            },
          },
        },
      }

      return opts
    end,
  },
}
```
I got this xml file from [Google Style Guide](https://github.com/google/styleguide/blob/gh-pages/eclipse-java-google-style.xml) and configured it locally to format with 4 spaces

**Reply by idelice** [reply]
Thanks but how do I use a command instead? E.g. I use mvn fmt:format to format my code.

**Reply by rogeriofrsouza** [reply]
conform.nvim has some [configuration options](https://github.com/stevearc/conform.nvim?tab=readme-ov-file#options) to set custom commands.
Could you try something like this?

```lua
return {
  "stevearc/conform.nvim",
  opts = {
    formatters = {
      my_formatter = {
        command = "mvn fmt:format",
      }
    }
  }
}
```

**Reply by spacealex69** [reply]
@rogeriofrsouza does your method skill work for you? i tried (with fresh lazyvim) your jdtls config example with change in xml style file `<setting id="org.eclipse.jdt.core.formatter.tabulation.size" value="4"/>` and still trying to figure out why profile not applied ...

**Reply by rogeriofrsouza** [reply]
Hey @spacealex69 👋🏻 

I updated my format settings not long after that answer, I  ended up using the [default jdtls formatter](https://github.com/rogeriofrsouza/dotfiles/blob/06836e8121d0fe9658458556d1da24ec14d9ba2a/dot_config/nvim/lua/plugins/nvim-jdtls.lua#L81) with some [settings in a file](https://github.com/rogeriofrsouza/dotfiles/blob/main/jdt-formatter.prefs) and a `.editorconfig` to set the 4 space indentation.

But configuring a formatter like this is ultra sad, the easiest option is still using that Google formatter, if you need 4 space just include `--aosp` arg in [conform.nvim file](https://github.com/rogeriofrsouza/dotfiles/blob/06836e8121d0fe9658458556d1da24ec14d9ba2a/dot_config/nvim/lua/plugins/conform.lua#L7).

**Reply by spacealex69** [reply]
@rogeriofrsouza thanks for sharing! will try your approach <3

**Comment by outwittester** [comment]
if you have google-java-format installed:
put below into your conform.vim setup will overwrite google-java-format default 2 indent
```
conform.setup({
 	formatters = {
		["google-java-format"] = {
			prepend_args = { "--aosp" }, -- Ensure 4-space indentation
		},
	},
})
```

**Reply by TungNguyen-dev** [reply]
Very helpful, thankyou

**Reply by Zombiesalad1337** [reply]
corresponding lua version:

```
require("conform").setup({
  formatters = {
    ["google-java-format"] = {
      prepend_args = { "--aosp" }, -- ensures 4-space indentation
    },
  },
})
```

Saved my day, thanks!
